pkg/monitors/elasticsearch/query/monitor.go

```diff
@@ -0,0 +1,136 @@
+}
+
+// Conceptually the data field in search queries define  bucket aggregations
+// that determine "buckets" (some group of elasticsearch documents that match
+// the search query) and metric aggregations that compute metrics from each "bucket"
+type metricQuery struct {
+	// Index that's being queried
+	Index string `yaml:"index" validate:"required"`
+	// Lucene search query which the elasticsearch documents should match
+	LuceneSearchQuery string `yaml:"luceneSearchQuery" validate:"required"`
+	// Query Interval
+	QueryInterval time.Duration `yaml:"queryInterval" default:"30s"`
+	// HTTP Post data field for aggregations
+	HTTPPostData string `yaml:"httpPostData" validate:"required"`
+}
+
+// Monitor for ES queries
+type Monitor struct {
+	Output types.FilteringOutput
+	cancel context.CancelFunc
+	ctx    context.Context
+	logger *utils.ThrottledLogger
+}
+
+func init() {
+	monitors.Register(&monitorMetadata, func() interface{} { return &Monitor{} }, &Config{})
+}
+
+type metricQueryInfo struct {
+	interval float64
+	aggsMeta map[string]*AggregationMeta
+}
+
+// Configure monitor
+func (m *Monitor) Configure(c *Config) error {
+	m.logger = utils.NewThrottledLogger(log.WithFields(log.Fields{"monitorType": monitorType}), 20*time.Second)
+
+	httpClient, err := c.HTTPConfig.Build()
+	if err != nil {
+		return err
+	}
+	esClient := NewESQueryClient(c.Host, c.Port, c.HTTPConfig.Scheme(), httpClient)
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+
+	metricQueryInfos := make([]*metricQueryInfo, len(c.MetricQueries))
+
+	for i, mq := range c.MetricQueries {
+		var reqBody HTTPPostData
+		err = json.Unmarshal([]byte(mq.HTTPPostData), &reqBody)
+		if err != nil {
+			return err
+		}
+
+		aggsMeta, err := reqBody.getAggregationsMeta()
+		if err != nil {
+			return err
+		}
+
+		metricQueryInfos[i] = &metricQueryInfo{
+			interval: mq.QueryInterval.Seconds(),
+			aggsMeta: aggsMeta,
+		}
+	}
+
+	// Since metric queries can have their own interval, set them running on separate go routines
+	for i, mq := range c.MetricQueries {
+		index := i
+		query := mq
+		utils.RunOnInterval(m.ctx, func() {
+			body, err := esClient.makeHTTPRequestFromConfig(query)
+
+			if err != nil {
+				log.Errorf("Failed to make HTTP request: %s", err)
+				return
+			}
+
+			var resBody HTTPResponse
+			if err := json.Unmarshal(body, &resBody); err != nil {
+				log.Errorf("Error processing HTTP response: %s", err)
+				return
+			}
+
+			dps := collectDatapoints(resBody, metricQueryInfos[index].aggsMeta, map[string]string{
+				"index": c.MetricQueries[index].Index,
+			})
+
+			m.sendDatapoints(dps)
+		}, time.Duration(metricQueryInfos[i].interval)*time.Second)
+	}
+	return nil
+}
+
+func (m *Monitor) sendDatapoints(dps []*datapoint.Datapoint) {
+	// This is the filtering in place trick from https://github.com/golang/go/wiki/SliceTricks#filter-in-place
+	n := 0
+	for i := range dps {
+		if dps[i] == nil {
```

Comment:
You should really fix the thing that is inserting nil datapoints so that it doesn't send nil datapoints.  Then you can get rid of this method and just use `m.Output.SendDatapoints` directly.

Reply:
I don't think there'll be nil values, I'll make sure and remove this logic from here